src/envoy/http/path_rewrite/filter.cc

```diff
@@ -0,0 +1,119 @@
+  }
+
+  absl::string_view original_path = headers.Path()->value().getStringView();
+  // Reject requests with fragment identifiers. They should never be sent to
+  // servers, and it breaks how we handle path translation (query params
+  // appended incorrectly).
+  if (absl::StrContains(original_path, "#")) {
+    config_->stats().denied_by_invalid_path_.inc();
+    rejectRequest(
+        Envoy::Http::Code::BadRequest,
+        "Path cannot contain fragment identifier (#)",
+        utils::generateRcDetails(utils::kRcDetailFilterPathRewrite,
+                                 utils::kRcDetailErrorTypeBadRequest,
+                                 utils::kRcDetailErrorFragmentIdentifier));
+    return FilterHeadersStatus::StopIteration;
+  }
+
+  // Make sure route is calculated
+  auto route = decoder_callbacks_->route();
+  if (route == nullptr || route->routeEntry() == nullptr) {
+    config_->stats().denied_by_no_route_.inc();
+
+    rejectRequest(
+        Envoy::Http::Code::NotFound,
+        absl::StrCat("Request `", utils::readHeaderEntry(headers.Method()), " ",
+                     utils::readHeaderEntry(headers.Path()),
+                     "` is not defined by this API."),
+        utils::generateRcDetails(utils::kRcDetailFilterPathRewrite,
+                                 utils::kRcDetailErrorTypeUndefinedRequest));
+    return FilterHeadersStatus::StopIteration;
+  }
+
+  const auto* per_route =
+      route->routeEntry()->perFilterConfigTyped<PerRouteFilterConfig>(
+          kFilterName);
+  if (per_route == nullptr) {
+    ENVOY_LOG(debug, "no per-route path_rewrite config");
+    config_->stats().path_not_changed_.inc();
+    return FilterHeadersStatus::Continue;
+  }
+
+  std::string new_path;
+  if (!per_route->config_parser().rewrite(original_path, new_path)) {
+    config_->stats().denied_by_url_template_mismatch_.inc();
+    rejectRequest(
+        Envoy::Http::Code::InternalServerError,
+        absl::StrCat("Request `", utils::readHeaderEntry(headers.Method()), " ",
+                     utils::readHeaderEntry(headers.Path()),
+                     "` is mismatched with url_template."),
+        utils::generateRcDetails(utils::kRcDetailFilterPathRewrite,
+                                 utils::kRcDetailErrorTypeUndefinedRequest));
+    return FilterHeadersStatus::StopIteration;
+  }
+
+  config_->stats().path_changed_.inc();
+  if (!headers.EnvoyOriginalPath()) {
+    headers.setEnvoyOriginalPath(headers.getPathValue());
+  }
+  headers.setPath(new_path);
+  return FilterHeadersStatus::Continue;
+}
+
+void Filter::rejectRequest(Envoy::Http::Code code, absl::string_view error_msg,
+                           absl::string_view details) {
+  ENVOY_LOG(debug, "{}", error_msg);
+  decoder_callbacks_->sendLocalReply(code, error_msg, nullptr, absl::nullopt,
+                                     details);
+  decoder_callbacks_->streamInfo().setResponseFlag(
+      Envoy::StreamInfo::ResponseFlag::UnauthorizedExternalService);
```

Comment:
Better response flag we can use? This doesn't make much sense

Reply:
I could not find a better flag than this except ResponseFlag::NoRouteFound for not-route case. My take is we (service control, path-rewrite, and backend-auth) are external sever (for envoy) since we could not define our own flag

Reply:
OK, for backend-auth and path-rewrite, I changed it to.  not call setResponseFlag for most error, except NotRouteFound.